# FEATURE - unresolved block pool levelDB backup

MoonSeongKwang opened the pull request:
Unresolved block pool에서의 백업/복구를 구현하였습니다.

블록 자체에 대한 정보는 처음 block_msg가 올 때 그 json을 to_cbor 하여 저장합니다.
SCE를 거친 result query로 인한 결과(resolve 되기 전의 해당 block의 임시 state 값)는 이전 연결 블록들과 DB를 조회한 후 계산되어 나오는 값이기 때문에, 일반적인 구조체로 되어있습니다.
하지만 구조체는 levelDB에서 value로써 저장할 수가 없으므로, json 형태로 변환하여 to_cbor 하여 저장합니다.
user ledger list, contract ledger list, user attribute list, user cert list, contract list의 다섯 가지 임시값이 존재하는데, 이들은 array형 json으로 변환됩니다.
백업은 base58 형태의 block_id를 key로 하여 각 디렉토리에 저장됩니다.

복구인 restorePool은 현재로서는 프로그램이 처음 가동되면 호출되며, leveldb에 저장된 값들을 from_cbor하여 다시 구조체로 파싱하고 원상태로 저장합니다.

+ **추가 언급사항**: 여기가 아니라 이전인 feature/result-query에서 언급했어야 하는 사항인데, result query에 의한 처리 결과가 map으로 저장되기 때문에 각 ledger가 처리되는 순서와 반복문으로 접근하는 순서가 서로 다를 수 있습니다. (예시 : result는 B 잔액변경, A 잔액변경, C 잔액변경의 순서인데 처리된 결과값을 반복문으로 순회하면 A B C 순서로 출력)
하지만 result query는 처리 순서에 따라 잔액 등이 달라져서 처리의 여부가 바뀔 수 있지만, 이 result query를 올바르게 처리했다면 결과인 state들은 중복되는 관계가 없습니다. 그러므로 백업 등에서 state를  어떤 순서로 처리하건 문제가 생기지 않습니다.

MoonSeongKwang committed:
# ADD - block info backup
block의 backup을 구현하였습니다.
result query에 의한 값들의 백업은 뒤이어 구현될 예정입니다.

MoonSeongKwang committed:
# ADD - Results backup
result query에 의하여 unresolved block에 저장되는 미확정 상태의 ledger와
각종 정보들을 백업하는 기능을 구현하였습니다.
각 구조체들의 요소들을 json array 형식으로 삽입한 후, cbor로
serialize하여 저장합니다.

MoonSeongKwang committed:
# FEATURE - UBP backup save
백업 중 저장하는 기능에 관한 사항들을 추가하였습니다.
어떤 블록들이 백업되어 있는지를 기록하는 UNRESOLVED_BLOCK_IDS를 추가하여
관리합니다. resolve되거나 선택받지 못하여 pool에서 삭제되는 블록은
백업에서도 더이상 존재할 필요가 없으므로 백업 목록에서 삭제되고,
백업에서도 지워집니다.

MoonSeongKwang committed:
Merge branch 'develop' into feature/unresolved-block-pool-kvbackup

MoonSeongKwang committed:
# FEATURE - UBP backup load
백업을 leveldb에서 불러오는 기능을 구현하였습니다.
각종 예외처리는 TODO로 남겨둔 상태입니다.
json의 to_cbor로 저장되어있던 블록 정보 / result 정보들을 가져와서
from_cbor처리를 한 후, 반복문으로 다시 파싱해서 저장합니다.

또, chain->pushBlock에서 is_store 인자가 삭제되었습니다.
기존에는 chain->pushBlock 함수 내부에서 블록 정보에 대한 백업을 실시했기 때문에
이 인자를 사용하였지만, 현재는 block_msg를 추가로 변환할 필요 없이 처음
받았을 때 바로 백업에 저장하기 위하여 pushBlock과 saveBackupBlock을
분리하였기 때문에 해당 인자가 필요 없어졌습니다.

morishjs commented on src/plugins/chain_plugin/chain.cpp:
```diff
@@ -216,6 +231,123 @@ string Chain::getValueByKey(string what, const string &base_keys) {
+
+    nlohmann::json block_msg = nlohmann::json::from_cbor(kv_controller->loadBackupBlock(block_id));
+    Block restored_block;
+    restored_block.initialize(block_msg);
```
initialize 가 실패하는 case도 처리가 필요해보입니다.

morishjs commented on src/plugins/chain_plugin/chain.cpp:
```diff
@@ -216,6 +231,123 @@ string Chain::getValueByKey(string what, const string &base_keys) {
+  for (auto &each_block_id : id_array_json) {
+    base58_type block_id = each_block_id;
+
+    nlohmann::json block_msg = nlohmann::json::from_cbor(kv_controller->loadBackupBlock(block_id));
```
serialized_id_array 가 empty string인 경우가 있습니다.
`from_cbor`에서 exception이 발생하면 어떻게 처리하나요?

morishjs commented on src/plugins/chain_plugin/chain.cpp:
```diff
@@ -204,8 +204,23 @@ void Chain::saveChain(local_chain_type &chain_info) {
+void Chain::saveBlockIds() {
+  nlohmann::json id_array = unresolved_block_pool->getPoolBlockIds();
+
+  kv_controller->saveBlockIds(TypeConverter::bytesToString(nlohmann::json::to_cbor(id_array)));
```
kv_con->saveBlockIds의 파라미터를 string->bytes로 변경하는게 낫지 않을까요?
어짜피 json::cbor 가 byte로 리턴되니까요. 다른 곳에서도 항상 json->bytes->string 으로 타입 변경을 해야해서 불편할 것 같습니다.

MoonSeongKwang replied:
levelDB의 value값이 기본적으로 string형이어야 해서, 저장할 때 string으로 저장하려고 늘 변환합니다. block ids들도 결국 array형태로 유지되어서 levelDB에 저장되는 값이라서요.

morishjs replied:
@SeongKwangMoon 
제 말은 saveBlockIds(string) -> saveBlockIds(bytes) 파라미터 타입을 변경해서
saveBlockIds 함수 implementation 에서 bytes->string으로 변환하면 다른 곳에서 여러번 타입을 변경하는 일을 줄일 수 있지 않을까 였습니다.

MoonSeongKwang replied:
@morishjs 
무슨 말씀을 하시는지는 이해했는데, 타입 변경은 `Chain::saveBlockIds`에서 `KvCon::saveBlockIds`으로 전달할 때 한 번만 이루어집니다. 함수 특성상 `Chain::saveBlockIds`가 아닌 다른 곳에서 `KvCon::saveBlockIds`를 호출할 일도 없구요.

지금 kv_store.cpp는 따로 타입 변경하는 일 없이 전달받은 값을 그대로 저장만 하도록 구현해둔거라서 이렇게 했습니다.

말씀하신 사항 보면서 살펴보니, 오히려 그 아래에 있는 KvCon::saveBackupBlock이 다른 함수들과 다르게 kv_store.cpp에서 타입 변경을 하네요. 함수 구현의 통일성을 유지하려면 이부분을 수정해야 할 것 같습니다.

또는 아예 '타입 변경 후 저장' 까지를 kv_store.cpp의 구현 방향으로 통일할 수도 있긴 합니다. 지금으로썬 당장 어느쪽이 더 효율적인지 판단이 안 서네요.

morishjs replied:
어느 쪽이 더 효율적인지 모를때는 그냥 이대로 두고 나중에 수정하면 될 것 같습니다.

morishjs commented on src/plugins/chain_plugin/kv_store.cpp:
```diff
@@ -188,6 +178,97 @@ void KvController::destroyDB() {
+  commitBatchAll();
+}
+
+bool KvController::saveBackupBlock(const nlohmann::json &block_json) {
```
saveBlock 과 saveBackupBlock 가 서로 다른 데이터를 저장하는건가요?
saveBackupBlock 의 구현내용을 보면 그냥 블록을 serialize 해서 저장하는거라서 backup이라는 특별한 의미가 없어보이네요.

MoonSeongKwang replied:
saveBlock 이라는 함수는 없고, 만약 saveBlockIds를 말씀하시는 거라면 둘은 다릅니다.
```
struct UnresolvedBlock {
  Block block;
  int32_t cur_vec_idx{-1};
  int32_t prev_vec_idx{-1};
  int32_t ssig_sum{0};
  std::map<string, user_ledger_type> user_ledger_list;
  std::map<string, contract_ledger_type> contract_ledger_list;
  std::map<base58_type, user_attribute_type> user_attribute_list;
  std::map<base58_type, user_cert_type> user_cert_list;
  std::map<contract_id_type, contract_type> contract_list;
```

Unresolved block은 msg_block으로부터 받은 블록의 기본 정보들(block id, signers, 각종 sig 등등)을 저장하는 block과 SCE의 result query를 받아 처리한 결과들을 가집니다.
`Chain::saveBackupBlock`은 전자인 블록의 기본 정보를 저장하는 함수이며, 
`Chain::saveBackupResult`는 후자인 처리 결과들을 저장합니다.
그리고 block id를 key로 해서 levelDB에 저장하는데, 어떤 block id들이 저장되어있었는지를 알 길이 없으므로 현재 저장되어있는 block id들을 고정 경로에 저장하는 `saveBlockIds` 함수가 유지됩니다.

MoonSeongKwang replied:
backup은 말그대로 backup이라서 메모리에 있는 정보들을 disk인 levelDB에 저장해뒀다가 다시 실행시켰을 때 복구하기 위한 거라서, 그냥 수신한 msg_block을 serialize해서 저장하는게 맞습니다.
다시 실행되었을 때 주변 노드들에게 msg_block을 요청하게 만들 수도 있긴 하겠지만, 비효율적일 것 같네요.

morishjs replied:
지금은 saveBlock이 구현 안되어있지만 나중에 다른 노드로부터 전파되어오는 block 을 저장하는 메소드가 생기면, saveBackupBlock과 코드 내용이 완전히 같지 않을까요? 그럼 구분이 필요가 없을 것 같은데요.

MoonSeongKwang replied:
제 생각으로는 다른 노드에서 전파되는 블록은 결국 메모리 상으로 들어오는 데이터일거고, saveBackupBlock은 메모리에 있는걸 levelDB로 저장하는거니까 역할이 다르지 않나요?

말씀하신 사항으로 보면 msg_block으로 메모리에 블록을 추가하는 pushBlock과 역할이 같을 것 같은데, 제가 뭔가 다르게 이해하고 있는 건가요?

blackironj replied:
아마 제 생각이 맞다면, MSG_BLOCK 이 왔을 때는 block 이 save 되는 곳은` unresolved pool `이 될거 같네요. 해당 block이 연결 할 수 있는 것 인지 모르기 때문에요.

morishjs commented:
고생하셨어요.
왜 오래 걸리는 작업인지 알겠네요. 예를 들면, `UserLedgers` 나 `ContractLedgers` 의 필드들이 미묘하게 다른데 (육안상 2~3개?) UserLedger/ContractLedger 를 꼭 나눠야 하는 이유가 없다면 기획자한테 말해서 고생을 조금 덜 수 있었을 것 같아요

blackironj commented on src/plugins/chain_plugin/include/chain.hpp:
```diff
@@ -55,11 +55,19 @@ class Chain {
+  void restoreContractLedgerList(UnresolvedBlock &restored_unresolved_block, const nlohmann::json contract_ledger_list_json);
+  void restoreUserAttributeList(UnresolvedBlock &restored_unresolved_block, const nlohmann::json user_attribute_list_json);
+  void restoreUserCertList(UnresolvedBlock &restored_unresolved_block, const nlohmann::json user_cert_list_json);
+  void restoreContractList(UnresolvedBlock &restored_unresolved_block, const nlohmann::json contract_list_json);
```
위의 해당 함수들이 json 객체를 복사해서 넘겨주는데 혹시 복사해야하는 이유가 있나요?
만약, 처음에 시작시에만 사용하는것이아닌 추후 다른 부분에서도 사용하게 된다면 그리 좋아보이지는 않습니다.

MoonSeongKwang replied:
이건 제가 빠뜨린 부분이네요. 참조로 바꾸겠습니다.

SeongKwangMoon commented:
User State Tree와 Contract State Tree가 별개로 유지되도록 설계가 바뀌었기 때문에 어쩔 수 없었습니다.
그래서 코드도 실제보다 양도 많아졌구요.

morishjs commented:
>그리고 각종 예외처리는 일단 제쳐두고 중요한 구현부터 하라는 말씀이 있으셔서 예외처리가 많이 미흡합니다

@SeongKwangMoon 네 그래서 예외처리 코멘트는 resolved 처리했습니다. resolved 처리 안된 것들만 보시고 수정하면 될 것 같습니다.

MoonSeongKwang committed:
# FIX - saveBackupBlock, json &
saveBackupBlock을 chain쪽에서 타입변환하고 kv에서는 저장만 하도록
통일하였습니다.
restoreXXX 함수 인자의 json에 &가 빠져있는것을 수정하였습니다.

SeongKwangMoon commented:
json 인자 참조 수정했고, 예외처리 관련해서는 코드에 TODO로 주석 남겨뒀습니다.
그리고 saveBackupBlock을 chain쪽에서 타입변환하고 kv_con에서는 저장만 하도록 통일하였습니다.
